sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java

```diff
@@ -3,83 +3,402 @@
+        return ClientRegistration.withRegistrationId(registrationId)
+                                 .clientName(registrationId)
+                                 .authorizationGrantType(new AuthorizationGrantType(
+                                     AADAuthorizationGrantType.ON_BEHALF_OF.getValue()))
+                                 .scope(clientProperties.getScopes())
+                                 .redirectUri("{baseUrl}/login/oauth2/code/")
+                                 .userNameAttributeName(properties.getUserNameAttribute())
+                                 .clientId(toClientId(properties))
+                                 .clientSecret(properties.getClientSecret())
+                                 .authorizationUri(endpoints.authorizationEndpoint());
+    }
+
+    private ClientRegistration.Builder toClientCredentialsBuilder(String registrationId,
+                                                                  AADAuthorizationServerEndpoints endpoints,
+                                                                  AuthorizationClientProperties clientProperties) {
+        return ClientRegistration.withRegistrationId(registrationId)
+                                 .clientName(registrationId)
+                                 .authorizationGrantType(AuthorizationGrantType.CLIENT_CREDENTIALS)
+                                 .scope(toScopes(clientProperties))
+                                 .redirectUri(properties.getRedirectUriTemplate())
+                                 .userNameAttributeName(properties.getUserNameAttribute())
+                                 .clientId(toClientId(properties))
+                                 .clientSecret(properties.getClientSecret())
+                                 .tokenUri(endpoints.tokenEndpoint());
+    }
+
+    private AzureClientRegistration emptyAzureClientRegistration() {
+        ClientRegistration.Builder clientRegistrationBuilder = toClientRegistrationBuilder(
+            AZURE_CLIENT_REGISTRATION_ID, defaultAzureAuthorizationClientProperties());
+        return new AzureClientRegistration(clientRegistrationBuilder.build(), new HashSet<>());
+    }
+
+    private AuthorizationClientProperties defaultAzureAuthorizationClientProperties() {
+        AuthorizationClientProperties result = new AuthorizationClientProperties();
+        result.setAuthorizationGrantType(AUTHORIZATION_CODE);
+        return result;
+    }
 
-        allClients = new HashMap<>();
-        addClientRegistration(azureClient.getClient());
-        for (ClientRegistration c : otherClients) {
-            addClientRegistration(c);
+    private String toClientId(AADAuthenticationProperties properties) {
+        return Optional.of(properties)
+                       .map(AADAuthenticationProperties::getClientId)
+                       .filter(StringUtils::hasText)
+                       .orElse("client-id-not-configured");
```

Comment:
what is this?

Reply:
It can be removed, but the exception will be more readable when the client id is missing.

Reply:
Move this logic to `AADAuthenticationProperties#afterPropertiesSet`.

Reply:
Ok

Reply:
It's required when the resource server with obo enabled, it depends on the application type, so we keep it here.